fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/ComparatorCursorContinuation.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * ComparatorCursorContinuation.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2021 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.cursors;
+
+import com.apple.foundationdb.record.ByteArrayContinuation;
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.RecordCursorEndContinuation;
+import com.apple.foundationdb.record.RecordCursorProto;
+import com.apple.foundationdb.record.RecordCursorStartContinuation;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Collections;
+import java.util.List;
+
+class ComparatorCursorContinuation extends MergeCursorContinuation<RecordCursorProto.ComparatorContinuation.Builder, RecordCursorContinuation> {
```

Comment:
What is the semantics for different advances among the plans? Reference plan advances by 5 records, challenger advances by 10: What do we do now? I would assumer to return the 5 records from the reference and _reset_ the other continuation to that record as well? Not really sure how this works.

Reply:
Since the goal was to log an alert if the plans do not return the same records, I took the approach of allowing the cursors to advance in unison  (meaning that it consumes one record from each when they all have a `next` result). Of course, this means that if a contender "misses" a result, it will never get back in sync with the reference cursor, but from an overall goal perspective, this is OK since the comparison already failed.
It would be more complicated to try and "resync" the cursors after a comparison failure, and I believe it will not add much value, since we would have to resolve the initial comparison failure anyways.

Reply:
Also: All the plans advance at the same rate - one record at a time - and the iteration stops when one of them reaches a limit. If this limit is "allowed" (e.g. scan limit or time limit), then wee stop the iteration and return a continuation. If the limit is semantically wrong (w/comparison to the reference plan - e.g. record limit or exhaustion) - then we log an error.
In all cases, the reference plan determines the record returned, and the termination status (e.g. exhaustion or record limit).

Reply:
What I meant is that we have seen plans hit the fetch limit. Imagine plan A being an index plan and plan B being a regular primary scan. In that case, one plan (the regular scan) may hit the scan limit by itself while the other one does not. What I think you are writing is that the reference plan decides how we advance and we return the continuations for both sides accordingly.

Related to this: I think the two plans now shared the scan limit, we should be cognizant of that fact.

Reply:
I believe this is handled correctly now: After the changes to the cursor, the returned result is the "strongest" of the results of the sub-plans, which means that the inefficient plan may spend more of the scan budget, and which will cause an "put-of-band" limit to be hit (either for itself or for the other plan). In that case the wrapping cursor will return that same out-of-band error and cause the plan execution to stop and return a continuation. This means that when this limit is hit, the scanning will continue until there is a result returned from all sub-cursors, at which point the comparison will take place.